src/core/models.py

```diff
@@ -20,26 +28,46 @@ class StateBase:
     """Base state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATES
+        self,
+        relation: Relation | None,
+        data_interface: DataRelation,
+        component: Unit | Application,
```

Comment:
Gotta check: do we really need `component` still...?

Reply:
Yeah I think so. In `ZKServer` which inherits from that, we use `component` in deciding `hosts`.